pkg/telemetry/client.go

```diff
@@ -0,0 +1,274 @@
+	if tc.installedPluginsByTargetType == nil {
+		tc.installedPluginsByTargetType = make(map[configtypes.Target]map[string]*cli.PluginInfo, 1)
+	}
+
+	if tc.installedPluginsByTargetType[target] == nil {
+		m := make(map[string]*cli.PluginInfo, 1)
+		tc.installedPluginsByTargetType[target] = m
+	}
+
+	tc.installedPluginsByTargetType[target][plugin.Name] = plugin
+}
+
+func (tc *telemetryClient) UpdateCoreCommandMap(coreCMDs ...*cobra.Command) {
+	if tc.coreCommandMap == nil {
+		tc.coreCommandMap = make(map[string]bool, 1)
+	}
+
+	for _, cmd := range coreCMDs {
+		tc.coreCommandMap[cmd.Name()] = true
+	}
+}
+
+func (tc *telemetryClient) UpdateCmdPreRunMetrics(cmd *cobra.Command, args []string) error {
+	if err := ensureMetricsSource(); err != nil {
+		return errors.Wrap(err, "failed to ensure metrics source in the configuration file")
+	}
+
+	cliID, err := cliInstanceID()
+	if err != nil {
+		return errors.Wrap(err, "unable to get CLI Instance ID")
+	}
+
+	if tc.isCoreCommand(cmd.CommandPath()) {
+		return tc.updateMetricsForCoreCommand(cmd, args, cliID)
+	}
+
+	return tc.updateMetricsForPlugin(cmd, args, cliID)
+}
+
+func (tc *telemetryClient) UpdateCmdPostRunMetrics(metrics *PostRunMetrics) error {
+	if metrics == nil {
+		return errors.New("post metrics data is required for update")
+	}
+	tc.currentOperationMetrics.ExitStatus = metrics.ExitCode
+	tc.currentOperationMetrics.EndTime = time.Now()
+	return nil
+}
+
+func (tc *telemetryClient) SaveMetrics() error {
+	// If cli command fail cobra validation, the PersistentPreRunE() wouldn't be invoked where initialization is done
+	// so, it is safe to ignore the metrics for user errors(like typos) at least to an extent where cobra can validate.
+	if tc.currentOperationMetrics.StartTime.IsZero() {
+		return nil
+	}
+
+	err := tc.metricsDB.CreateSchema()
+	if err != nil {
+		return errors.Wrap(err, "unable to create the telemetry schema")
+	}
+
+	return tc.metricsDB.SaveOperationMetric(tc.currentOperationMetrics)
+}
+
+// SendMetrics sends the local stored metrics to super collider
+// TODO: to be implemented
+func (tc *telemetryClient) SendMetrics() error {
+	return nil
+}
+
+func (tc *telemetryClient) isCoreCommand(cmdPath string) bool {
+	if cmdPath != "" {
+		cmds := strings.Split(cmdPath, " ")
+		if len(cmds) >= 2 && tc.coreCommandMap[cmds[1]] {
+			return true
+		}
+	}
+
+	return false
+}
+func (tc *telemetryClient) updateMetricsForCoreCommand(cmd *cobra.Command, args []string, cliID string) error {
+	tc.currentOperationMetrics.CliID = cliID
+	tc.currentOperationMetrics.CliVersion = buildinfo.Version
+	tc.currentOperationMetrics.StartTime = time.Now()
+	tc.currentOperationMetrics.CommandName = strings.Join(strings.Split(cmd.CommandPath(), " ")[1:], " ")
+
+	// CLI recommendation is to have a single argument for a command
+	if len(args) != 0 {
+		tc.currentOperationMetrics.NameArg = hashString(args[0])
+	}
```

Comment:
I remember something regarding why only a single argument is recommended got discussed somewhere, but can you point me to that discussion or can you remind me the reasons behind this as I am not able to remember now?

Reply:
https://github.com/vmware-tanzu/tanzu-cli/blob/main/docs/plugindev/style_guide.md#positional-arguments